pkg/gadgets/dns/tracer/bpf/dns.c

```diff
@@ -0,0 +1,120 @@
+// SPDX-License-Identifier: GPL-2.0
+/* Copyright (c) 2021 The Inspektor Gadget authors */
+
+#include <linux/bpf.h>
+#include <linux/if_ether.h>
+#include <linux/ip.h>
+#include <linux/in.h>
+#include <linux/udp.h>
+
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_endian.h>
+
+#include "dns-common.h"
+
+#define DNS_OFF (ETH_HLEN + sizeof(struct iphdr) + sizeof(struct udphdr))
+
+/* llvm builtin functions that eBPF C program may use to
+ * emit BPF_LD_ABS and BPF_LD_IND instructions
+ */
+unsigned long long load_byte(void *skb,
+			     unsigned long long off) asm("llvm.bpf.load.byte");
+unsigned long long load_half(void *skb,
+			     unsigned long long off) asm("llvm.bpf.load.half");
+unsigned long long load_word(void *skb,
+			     unsigned long long off) asm("llvm.bpf.load.word");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+} events SEC(".maps");
+
+// https://datatracker.ietf.org/doc/html/rfc1035#section-4.1.1
+union dnsflags {
+	struct {
+		__u8 rd :1;	// recursion desired
+		__u8 tc :1;	// truncation
+		__u8 aa :1;	// authoritive answer
+		__u8 opcode :4;	// kind of query
+		__u8 qr :1;	// 0=query; 1=response
+
+		__u8 rcode :4;	// response code
+		__u8 z :3;	// reserved
+		__u8 ra :1;	// recursion available
+	};
+	__u16 flags;
+};
+
+struct dnshdr {
+	__u16 id;
+
+	union dnsflags flags;
+
+	__u16 qdcount; // number of question entries
+	__u16 ancount; // number of answer entries
+	__u16 nscount; // number of authority records
+	__u16 arcount; // number of additional records
+};
+
+SEC("socket1")
+int bpf_prog1(struct __sk_buff *skb)
+{
+	// Skip non-IP packets
+	if (load_half(skb, offsetof(struct ethhdr, h_proto)) != ETH_P_IP)
+		return 0;
+
+	// Skip non-UDP packets
+	if (load_byte(skb, ETH_HLEN + offsetof(struct iphdr, protocol)) != IPPROTO_UDP)
```

Comment:
Isn't possible to make this BPF program triggered only for IP/UDP packets on port `53`?

Reply:
It's possible but we would miss DNS packets on alternative ports. OpenShift seems to use 5353 sometimes:
https://github.com/openshift/installer/issues/2291

I was hoping that the filter on the flags and qrcount/ancount would remove enough non-DNS UDP traffic.

Reply:
OpenShift is making things harder by following such a non-conventional configuration. 

I think that, as for `ancount`, you could add also a check on `nscount` given that it should determine the number of name server resource records in a DNS response. 

Let's test it, in case, we could consider adding for examples filters for only IPv4 and IPv6 requests (`A` and `AAAA` DNS query `type`) or only standard queries (`QUERY opcode`). Let's see.